oem/ibm/pfw-sms-utils/pfw_sms_menu.cpp

```diff
@@ -0,0 +1,280 @@
+                        detail.starts_with(
+                            "The password contains monotonic sequence longer than "))
+                    {
+                        pfw_sms_appdata_ptr->reasonCode =
+                            BAD_PASSWORD_TOO_LONG_MONOTONIC_SEQUENCE;
+                    }
+                    if (detail == "No password supplied")
+                    {
+                        pfw_sms_appdata_ptr->reasonCode =
+                            BAD_PASSWORD_NO_PASSWORD_SUPPLIED;
+                    }
+                    if (detail.starts_with(
+                            "The password fails the dictionary check"))
+                    {
+                        pfw_sms_appdata_ptr->reasonCode =
+                            BAD_PASSWORD_DICTIONARY_CHECK;
+                    }
+                }
+                else if (msg.starts_with("Username ") &&
+                         (std::string_view::npos != msg.find(" / Password ")) &&
+                         (std::string_view::npos !=
+                          msg.find(" exceeds IPMI 16/20 limit")))
+                {
+                    // Handle messages from
+                    // https://github.com/openbmc/pam-ipmi/blob/master/src/pam_ipmicheck/pam_ipmicheck.c
+                    pfw_sms_appdata_ptr->reasonCode =
+                        PASSWORD_TOO_LONG_FOR_IPMI_USER;
+                }
+                // Ignore other messges, such as: You are required to change
+                // your password immediately (administrator enforced).
+                break;
+            default:
+                // Should never get here
+                return PAM_CONV_ERR;
+                break;
+        }
+    }
+    return PAM_SUCCESS;
+}
+
+/** @brief Common code factored out between authenticate() and changePassword().
+ */
+static void common_authentication_handler(const std::string& username,
+                                          const std::string& password,
+                                          bool& authenticated,
+                                          bool& passwordChangeRequired,
+                                          bool& operationAllowed,
+                                          bool& passwordChangeAllowed)
+{
+    authenticated = false; // default to fail safe (no access)
+    passwordChangeRequired = false;
+    operationAllowed = false;
+    passwordChangeAllowed = false;
+
+    struct pfw_sms_pam_appdata appdata = {
+        password.c_str(), PASSWORD_CHANGE_FAILED_UNKNOWN_REASON};
+    char* appData = reinterpret_cast<char*>(&appdata);
+    const struct pam_conv pamConversation = {pamConversationFunction, appData};
+    pam_handle_t* pamHandle = nullptr; // this gets set by pam_start
+    int retval = pam_start("ibm-pfw-sms-menu", username.c_str(),
+                           &pamConversation, &pamHandle);
+    if (retval != PAM_SUCCESS)
+    {
```

Comment:
is pam_end missing here??
pam_end(pamHandle, retval); // ignore retval

Reply:
No.  I will add a comment to indicate that pam_start() failed, so no cleanup is needed.